README.md
# showballoon
showballoon is 32bit windows executable that shows a balloon (or tooltip).

## Command Line
If launched without command line options, it show a help message.

## Actual use
```
showballoon.exe /title:"My Title" /icon:"C:\Linkout\CommonExe\showballoon.exe" /defaulticon /duration:5000 /balloonicon:1 "This is a balloon message!"
```

## To pass non-ascii characters
You need to utf8-urlencode the non-ascii characters.
```
showballoon.exe /title:"%E7%A7%81%E3%81%AE%E3%82%BF%E3%82%A4%E3%83%88%E3%83%AB" /icon:"C:\Linkout\CommonExe\showballoon.exe" /defaulticon /duration:5000 /balloonicon:1 %E3%81%93%E3%82%8C%E3%81%AF%E3%83%90%E3%83%AB%E3%83%BC%E3%83%B3%E3%83%A1%E3%83%83%E3%82%BB%E3%83%BC%E3%82%B8%E3%81%A7%E3%81%99%EF%BC%81
```

```
showballoon.exe /title:"%E6%88%91%E7%9A%84%E9%A2%98%E7%9B%AE" /icon:"C:\Linkout\CommonExe\showballoon.exe" /defaulticon /duration:5000 /balloonicon:1 %E8%BF%99%E6%98%AF%E4%B8%AA%E6%B0%94%E7%90%83%E4%BF%A1%E6%81%AF!
```



History.txt
2023/04/25 ver2.0.1
Shows error if unknown options provided
Upgraded to VS2022

2021/06/24 ver1.5.1
Move to vs2019

2017/12/29 ver1.04
Start to use VS 2013

2017/10/14 ver1.03



showballoon.cpp
// showballoon.cpp :
//

#include "stdafx.h"


#include "../MyUtility/UTF16toUTF8.h"

#pragma comment(lib,"Comctl32.lib")

using namespace std;

HICON ghIcon;

BOOL NotifyIconize(UINT uID, DWORD dwMessage, LPCWSTR pInfoTitle , LPCWSTR pInfo)
{
	NOTIFYICONDATAW nid;
	ZeroMemory(&nid,sizeof(nid));
	nid.cbSize = sizeof(NOTIFYICONDATAW);
	nid.hWnd = NULL;
	nid.uID = uID;
	nid.uFlags = NIF_ICON | NIF_MESSAGE | NIF_TIP | 0x00000010;
	nid.dwInfoFlags      = 0x00000001;
	nid.uTimeout         = 300;
	nid.uCallbackMessage = 0; //WM_APP_TRAYMESSAGE;
	nid.hIcon = ghIcon; //LoadIcon(GetModuleHandle(NULL), MAKEINTRESOURCE(IDI_ICON_MAIN));
	lstrcpyW(nid.szTip, L"dater");
	if(pInfoTitle)
		lstrcpyW( nid.szInfoTitle, pInfoTitle );
	if(pInfo)
		lstrcpyW( nid.szInfo, pInfo );
	
	BOOL ret= Shell_NotifyIconW( dwMessage,&nid);
	//DestroyIcon(sfi.hIcon);
	return ret;
}

char a2c1(char c1)
{
	char ret=0;
	if('0'<=c1 && c1<='9')
	{
		ret=c1-'0';
	}
	else 
	{
		switch(c1)
		{
		case 'a': case 'A': ret=0xa; break;
		case 'b': case 'B': ret=0xb; break;
		case 'c': case 'C': ret=0xc; break;
		case 'd': case 'D': ret=0xd; break;
		case 'e': case 'E': ret=0xe; break;
		case 'f': case 'F': ret=0xf; break;
		}
	}
	return ret;
}

char a2c(char c1, char c2)
{
	return a2c1(c1)*16 + a2c1(c2);
}
		
wstring argToWstring(const char* p)
{
	string org;
	for(;*p;++p)
	{
		if(*p=='%')
		{
			if( *(p+1)==0 || *(p+2)==0)
				break;

			char c=a2c(*(p+1), *(p+2));
			org += c;
			p+=2;
		}
		else if(*p=='+')
		{
			org+=' ';
		}
		else
		{
			org+=*p;
		}
	}
	wstring ret;
	UTF8toUTF16((const LPBYTE)org.c_str(),ret);
	return ret;
}

int APIENTRY WinMain(HINSTANCE hInstance,
                     HINSTANCE hPrevInstance,
                     LPTSTR     lpCmdLine,
                     int       nCmdShow )
{
	
	if(__argc < 2)
	{
		wstring message;
		message += NS(L"Usage:\r\n");
		message += NS(L"showballoon.exe [/title:STRING] [/icon:RESOURCEFORICON] [/iconindex:i] STRING");
		message += L"\r\n";
		message += L"\r\n";
		message += L"STRING:\tUTF8 url encoded string";
		message += L"\r\n";
		message += L"RESOURCEFORICON:\tExe or Dll for finding icon";
		MessageBoxW(NULL, 
			message.c_str(),
			L"showballoon",
			MB_ICONINFORMATION);
		return 1;
	}

	wstring text;
	wstring title;
	wstring iconexe;
	int iconindex=0;
	UINT uTrayID=WM_APP+1;

	LPCSTR pTitleOption = "/title:";
	size_t nTitleOption = strlen(pTitleOption);
	
	LPCSTR pIconOption = "/icon:";
	size_t nIconOption = strlen(pIconOption);

	LPCSTR pIconIndexOption = "/iconindex:";
	size_t nIconIndexOption = strlen(pIconIndexOption);
	for(int i=1 ; i < __argc ; ++i)
	{
		if(0== strnicmp(__argv[i], pTitleOption, nTitleOption))
		{
			char* pS = __argv[i] + nTitleOption;
			title=argToWstring(pS);
		}
		else if(0== strnicmp(__argv[i], pIconOption, nIconOption))
		{
			char* pS = __argv[i] + nIconOption;
			iconexe=argToWstring(pS);
		}
		else if(0== strnicmp(__argv[i], pIconIndexOption, nIconIndexOption))
		{
			char* pS = __argv[i] + nIconIndexOption;
			iconindex = atoi(pS);
		}
		else
		{
			text = argToWstring(__argv[i]);
		}
	}

	CoInitialize(NULL);

	if(!iconexe.empty())
	{
		SHFILEINFOW sfi={0};
		SHGetFileInfoW(iconexe.c_str(),
					   0,
					   &sfi, 
					   sizeof(SHFILEINFO), 
					   // SHGFI_SYSICONINDEX|
					   SHGFI_ICON|
					   SHGFI_SMALLICON);


		ghIcon = sfi.hIcon;
	} 	
	

	InitCommonControls();
	HWND hBalloon = CreateWindowW(L"tooltips_class32",
		NULL,
		WS_POPUP|0x02|0x40|0x01,
		CW_USEDEFAULT,
		CW_USEDEFAULT,
		CW_USEDEFAULT,
		CW_USEDEFAULT,
		NULL,
		NULL,
		NULL,
		NULL);
	if(!hBalloon)
	{
		MessageBoxW(NULL, L"balloon",NULL,MB_ICONERROR);
		return __LINE__;
	}
	if(!SendMessageW(hBalloon,
		TTM_SETTITLE, // Adds a standard icon and title string to a ToolTip    
		1,
		(LPARAM)L"title"))
	{
		MessageBoxW(NULL, L"balloon sendmessage",NULL,MB_ICONERROR);
		return __LINE__;
	}

	NotifyIconize(uTrayID,NIM_DELETE, NULL, NULL);
	if(!NotifyIconize(uTrayID,NIM_ADD, NULL, NULL))
	{
		MessageBoxA(NULL, "NotifyAdd",NULL,MB_ICONERROR);
		return __LINE__;
	}



	if(!NotifyIconize(uTrayID, NIM_MODIFY, title.c_str(), text.c_str() ))
	{
		MessageBoxA(NULL, "NotifyModify",NULL,MB_ICONERROR);
		return __LINE__;
	}
	Sleep(5000);
	NotifyIconize(uTrayID, NIM_DELETE, NULL, NULL);
	DestroyWindow(hBalloon);
	DestroyIcon(ghIcon);
	return 0;
}





showballoonTest/main.cpp
//Copyright (C) 2017 Ambiesoft All rights reserved.
//
//Redistribution and use in source and binary forms, with or without
//modification, are permitted provided that the following conditions
//are met:
//1. Redistributions of source code must retain the above copyright
//notice, this list of conditions and the following disclaimer.
//2. Redistributions in binary form must reproduce the above copyright
//notice, this list of conditions and the following disclaimer in the
//documentation and/or other materials provided with the distribution.
//
//THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
//ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
//FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
//OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
//HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
//LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
//OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
//SUCH DAMAGE.


#include <Windows.h>
#include <tchar.h>
#include <shlwapi.h>

#include <string.h>
#include <string>

#include "../../lsMisc/OpenCommon.h"
#include "../../lsMisc/stdosd/stdosd.h"

#pragma comment(lib,"user32.lib")
#pragma comment(lib,"shlwapi.lib")

using namespace std;
using namespace Ambiesoft;
using namespace Ambiesoft::stdosd;


wstring strSB;
void testit(LPCTSTR pCommand)
{
	MessageBox(NULL,pCommand,_T("TEST"),MB_OK);
	OpenCommon(NULL, strSB.c_str(), pCommand);
}

int main()
{
	strSB = stdCombinePath(
		stdGetParentDirectory(stdGetModuleFileName()),
		L"showballoon.exe");
	if (!PathFileExists(strSB.c_str()))
	{
		MessageBox(NULL, L"NAI", 0, 0);
		return 1;
	}

	wstring command;
	
	command = L"ABC";
	testit(command.c_str());

	
	command = _T("%E3%81%88%E3%81%88%E3%81%84%E3%81%82%E3%81%82+%E5%90%9B%E3%81%8B%E3%82%89%E3%80%8C%E3%82%82%E3%82%89%E3%81%84%E6%B3%A3%E3%81%8D%E3%80%8D+%E3%81%BB%E3%82%8D%E3%82%8A%E3%83%BB%E3%81%BB%E3%82%8D%E3%82%8A+%E3%81%B5%E3%81%9F%E3%82%8A%E3%81%BC%E3%81%A3%E3%81%A1");
	testit(command.c_str());
	

	command = stdFormat(_T("/title:%s /icon:%s /defaulticon /balloonicon:%d /duration:%d %s"),
		_T("%E3%82%82%E3%82%89%E3%81%84%E6%B3%A3%E3%81%8D"), //title
		(wstring(_T("\"") + strSB + _T("\""))).c_str(), // icon
		1, // balloonicon
		60 * 1000,
		_T("%E3%81%88%E3%81%88%E3%81%84%E3%81%82%E3%81%82+%E5%90%9B%E3%81%8B%E3%82%89%E3%80%8C%E3%82%82%E3%82%89%E3%81%84%E6%B3%A3%E3%81%8D%E3%80%8D+%E3%81%BB%E3%82%8D%E3%82%8A%E3%83%BB%E3%81%BB%E3%82%8D%E3%82%8A+%E3%81%B5%E3%81%9F%E3%82%8A%E3%81%BC%E3%81%A3%E3%81%A1")
	);
	testit(command.c_str());


	command = stdFormat(_T("/title:%s /icon:%s /defaulticon /balloonicon:%d /duration:%d %s"),
		_T("%EC%98%81%EC%96%B4"), //title
		(wstring(_T("\"") + strSB + _T("\""))).c_str(), // icon
		1, // balloonicon
		60 * 1000,
		_T("%E8%8B%B1%E8%AF%AD%EF%BC%88English%EF%BC%8C%E8%81%86%E8%81%BDi%2F%CB%88%C9%AA%C5%8B%C9%A1l%C9%AA%CA%83%2F%EF%BC%89%E6%98%AF%E4%B8%80%E7%A7%8D%E8%A5%BF%E6%97%A5%E8%80%B3%E6%9B%BC%E8%AF%AD%E8%A8%80")
	);
	testit(command.c_str());

	command = L"/ttt:aaa This is an error";
	testit(command.c_str());
	return 0;
}